[git-webkit] Replace symbols with '-' when converting a bug title to a branch name

fujii opened the pull request:
#### 3f0d9a2960ccda0996b8b1622524cf107cae1f2e
<pre>
[git-webkit] Replace symbols with &apos;-&apos; when converting a bug title to a branch name
<a href="https://bugs.webkit.org/show_bug.cgi?id=283781">https://bugs.webkit.org/show_bug.cgi?id=283781</a>

Reviewed by Jonathan Bedard.

If a bug title contains &quot;REGRESSION(123@main):&quot;, `git-webkit branch`
just dropped symbols and converted it to a branch name
&quot;REGRESSION123main&quot;.

Replace consecutive symbols with &apos;-&apos; so that the above example yields
&quot;REGRESSION-123-main&quot;.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.to_branch_name):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:

Canonical link: <a href="https://commits.webkit.org/287228@main">https://commits.webkit.org/287228@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e74f8c7e8a27da1937729ed6b2639b7dc909cc49

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78696 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57740 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32078 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83356 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29958 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80829 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66892 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6021 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61632 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19557 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/70675 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41941 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/78206 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25637 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28298 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70116 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26017 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84725 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6061 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4182 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69859 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/78385 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6223 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67645 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69113 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13151 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11711 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12172 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6006 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5992 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7780 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
```diff
@@ -91,13 +91,9 @@ def editable(cls, branch, repository=None):
-        return string_utils.encode(result, target_type=str)
+        strings = re.split(r'\W+', string_utils.decode(value))
+        strings = [s for s in strings if s != ""]
+        return string_utils.encode('-'.join(strings), target_type=str)
```
Is there any reason not to use `re.sub(r'\w+', '-', string_utils.decode(value)).strip('-')`? Rather than the split/join logic, which took me a minute or two to understand (and having to check what exactly `re.split`'s behaviour is), I'd rather we just did that. (`string_utils.encode(..., target_type=str)` is totally redundant at this point, given `str` is always the Python 3 `str` now.)

That said, this also changes `to_branch_name('foo_bar')` from `foo_bar` to `foo-bar`.

fujii replied:
Good idea. Will revise.
We have a test case of `_`. We should keep `_`.

fujii replied:
Fixed.

gsnedders reviewed (changes requested):
Please see my comment below about `re.sub` + `str.strip`, because that would make this simpler to understand.

And, far as I can see, I don't think this is ever used accept when we're creating a new branch, so we can change the behaviour with no concerns about backwards compatibility, so changing this seems safe.

Also, mostly to note for future reference for anyone looking here in the future, looking at https://git-scm.com/docs/git-check-ref-format/2.43.0#_description, the format actually allowed is kinda complicated — while we can certainly be super-aggressive at removing symbols, we could also allow them sometimes.

fujii committed:
[git-webkit] Replace symbols with '-' when converting a bug title to a branch name
https://bugs.webkit.org/show_bug.cgi?id=283781

Reviewed by Jonathan Bedard.

If a bug title contains "REGRESSION(123@main):", `git-webkit branch`
just dropped symbols and converted it to a branch name
"REGRESSION123main".

Replace consecutive symbols with '-' so that the above example yields
"REGRESSION-123-main".

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/branch.py:
(Branch.to_branch_name):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/branch_unittest.py:

Canonical link: https://commits.webkit.org/287228@main
